### ISSUE TITLE
Loading data thru function (including server-side pagination/sort/filter)

### ISSUE THREAD

**Issue by sekarpdkt** [wenzhixin/bootstrap-table#5022]
Is there a way to use a function to load data instead of URL? This will be useful, if we are loading data thru other methodologies like websocket etc.

I have done the same by using an attribute source with following code changes. If required, can generate a PR. It worked for my use case as I used deferred to fetch the data in that called function. 

In `initServer` function changed

```
	      if (!(url || this.options.url ) && !this.options.ajax) {
	        return;
	      }
```
as

```
	      if (!(url || this.options.url || this.options.source) && !this.options.ajax) {
	        return;
	      }
```
Then added following block in same function **before** `var request=....`
```
          if(this.options.source){
            let _res=this.options.source(data);
	          var res = Utils.calculateObjectValue(_this10.options, _this10.options.responseHandler, [_res, undefined], _res);

	          _this10.load(res);

	          _this10.trigger('load-success', res, true, undefined);

	          if (!silent) {
	            _this10.hideLoading();
	          }

	          if (_this10.options.sidePagination === 'server' && res[_this10.options.totalField] > 0 && !res[_this10.options.dataField].length) {
	            _this10.updatePagination();
	          }            
            
            return data;
          }
```
I am able to get data from the function defined as `source:name-of-function`

Is there any other better option?

**Comment by wenzhixin** [comment]
Do you mean the [`data`](https://examples.bootstrap-table.com/#welcomes/from-data.html) option or the [`load`](https://examples.bootstrap-table.com/#methods/load.html) method?

**Comment by sekarpdkt** [comment]
Will load method calls same function when we need pagination / sort / refresh to be performed? In above approach same function is called whenever Ajax call was supposed to happen. (note: I did further minor tweaking. If we conclude it is a valid scenario, I can issue a PR).

As I mentioned, above will be useful, if you are planning to get data from multiple sources and combine it to show.

**Comment by wenzhixin** [comment]
Can you provide an [online example](https://live.bootstrap-table.com/) to show the feature?
```js
let _res=this.options.source(data);
var res = Utils.calculateObjectValue(_this10.options, _this10.options.responseHandler, [_res, undefined], _res);
```

It looks like this option has the same function as `responseHandler` option.

**Comment by sekarpdkt** [comment]
Sure.. Will give in few days. As in online option, I will not be able to call multiple wss, will try to find a way to simulate it.

**Comment by sekarpdkt** [comment]
Hi Unable to create an example,primarily because there is no option to call a function for data everytime without above mentioned change. I can link to a similar use case. In typeahead, I can configure multiple data sets for same typeahead dropdown. 

https://twitter.github.io/typeahead.js/examples/ and looks for Multiple Datasets.In this case it is able to merge data from two data source. I want to achieve something similar, which is working in my local implementation. In table config I added following changes.
`
      source:tableManager.data4Table,
`
And in bootstrap-table.js
```
          if(this.options.source){
                this.options.source(data,this.options).then((_res) =>{
                    var res = Utils.calculateObjectValue(_this10.options, _this10.options.responseHandler, [_res, undefined], _res);
                    _this10.load(res);
                    _this10.trigger('load-success', res, true, undefined);
                    if (!silent) {
                        _this10.hideLoading();
                    }
                    if (_this10.options.sidePagination === 'server' && res[_this10.options.totalField] > 0 && !res[_this10.options.dataField].length) {
                        _this10.updatePagination();
                    }            
                    return data;
                })
          }else{
/******************existing logic****************/
}

```
And function to get data
```
tableManager.data4Table =async (request,options) =>{
   /*
   Build query condition as payload
   */
    let dataFromSources=[]
    await Promise.all(dataSources.data.map(async (source) =>  {
           dataFromSources[source.id]=await getDataFromSource1(payload,source.config);
   }));

    //merge all source data as data
    return data;
}
```

Basically get the data from all sources for the given condition, merge them and send it as data for display.

**Comment by wenzhixin** [comment]
I think the [`ajax`](https://bootstrap-table.com/docs/api/table-options/#ajax) option can do what you want: https://live.bootstrap-table.com/code/wenzhixin/2904.

**Comment by sekarpdkt** [comment]
Hi.. Thanks. It solved original issue. I was having another request to send additional parameters to the AJAX function. This will be useful, if we  have multiple tables in the same page and we wanted to have a common AJAX handler. Is there any option available to pass customer additional info to ajax handler.

**Comment by sekarpdkt** [comment]
Ok. Solved it thru ` ajax:function(params) {ajaxhandler(additionalInfo,params); }`
